GC debugging aids

CyberShadow opened the pull request:
These tweaks were useful while debugging https://github.com/dlang-community/libdparse/pull/445.

They might as well be upstreamed.

CC @rainers

rainers commented on src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -1269,10 +1274,11 @@ struct Gcx
-                    &_d_gcx_atfork_prepare,
-                    &_d_gcx_atfork_parent,
-                    &_d_gcx_atfork_child);
+                version (COLLECT_PARALLEL)
```
Why are the fork handlers only needed when marking in parallel?

CyberShadow replied:
You tell me. `_d_gcx_atfork_prepare` etc. are inside the `version (COLLECT_PARALLEL):` section.

rainers replied:
Initially the fork handler only dealt with variables for parallel marking, but https://github.com/dlang/druntime/pull/2817 made it more general, so should have moved it out of the version.
It currently doesn't make much difference, though, as version COLLECT_PARALLEL is available unconditionally ATM.

CyberShadow replied:
Oh, I see. So, I screwed up then by adding these handlers in the wrong place.

Also I had completely forgotten that it was me who added them. :)

I will move them out.

`COLLECT_PARALLEL` is unconditional in that it is always turned on at the top of the file, but the point of such switches is to toggle them during debugging, so I think turning it off in this way ought to either work or the switch should just be removed (and the behavior controlled through the runtime setting).

CyberShadow replied:
> I will move them out.

Done.

rainers commented on src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -2464,19 +2470,25 @@ struct Gcx
+                        if (hasFinalizer)
+                            sweepBin!true();
+                        else
+                            sweepBin!false();
```
Does this compile to nothing if all debug options are off? I doubt that, so maybe prepend `debug`.

CyberShadow replied:
In DMD, it doesn't compile to nothing. I think it should compile to nothing in LDC/GDC but I haven't checked yet.

`debug` isn't quite right, there's a few conditions affecting this code.

rainers replied:
> In DMD, it doesn't quite compile to nothing.

I think that would be a blocker for adding the else path unconditionally. An appropriate compile time condition should be possible.

CyberShadow replied:
I suspect it would be inobservable in benchmarks, because the loop just increments registers.

But, I think it would be better to just split this into several loops, each with a top-level conditional.

CyberShadow replied:
> I think it should compile to nothing in LDC/GDC, but I haven't checked.

It does not:

https://forum.dlang.org/post/aejdlmkvuwgbtxktpjby@forum.dlang.org

CyberShadow replied:
> But, I think it would be better to just split this into several loops, each with a top-level conditional.

Done.

The best solution I could come up with here is a string mixin.

rainers commented on src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -4208,7 +4220,7 @@ debug (LOGGING)
 
 
-        private void log_free(void *p) nothrow @nogc
+        private void log_free(void *p, size_t size) nothrow @nogc
```
`size` unused? I suspect you added some additional checks in your local copy. Add it to the `printf`s here instead?

CyberShadow replied:
Please see the commit message.

This patch isn't quite right yet (as the failing tests show).

CyberShadow replied:
> Add it to the `printf`s here instead?

Good idea, done.

> This patch isn't quite right yet (as the failing tests show).

Fixed.

CyberShadow commented:
OK, I think this is good to go.

- Addressed suggestions from code review.
- Fixed some bitrot in old tests.
- Fixed and moved an old commented-out unittest fragment to its own test.
- Added a test for the debug hooks fix.

rainers commented on src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -213,10 +213,17 @@ class ConservativeGC : GC
         runLocked!(go, otherTime, numOthers)(gcx);
     }
 
+    debug (GC_RECURSIVE_LOCK) static bool lockedOnThisThread;
```
Please add the new condition to the list of conditions at the top with a short description.

BTW: does this change really help? IIRC throwing involves capturing the stack which is saved to memory allocated from the GC, so a staticError is not really `@nogc`. Maybe this has changed in the mean time, though.

CyberShadow replied:
Good idea, done.

Yep, it does seem to work! The tests I added make use of it to catch the `Error`.

CyberShadow replied:
> IIRC throwing involves capturing the stack which is saved to memory allocated from the GC

Yeah, `OutofMemoryError` errors do not have a stack trace, which incidentally makes them difficult to debug. Room for improvement there...

rainers commented on src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -2487,21 +2491,44 @@ struct Gcx
-                        }
+                        }`;
+
+                        debug (SENTINEL)
```
Please avoid the string mixin, it makes the code undebuggable.

I was rather thinking of:
```
debug(SENTINEL)
   enum doLoop = true;
else version(assert)
   enum doLoop = true;
else debug(...)
   ...
else
   bool doLoop = hasFinalizer;
if (doLoop)
````
followed by the existing code. If any of the debug conditions are enabled, peerformance is not a goal.

CyberShadow replied:
Undebuggable how? You mean with an interactive debugger?

I wanted to avoid duplicating the condition and putting the other instance too far away from the first one, as they will inevitably go out of sync, as they already have in this very case.

rainers replied:
> Undebuggable how? You mean with an interactive debugger?

Yes.

BTW: I just noticed that `hasFinalizer` is already a defined a few lines above as I proposed. Maybe it is just missing the SENTINEL condition?

CyberShadow replied:
> BTW: I just noticed that `hasFinalizer` is already defined a few lines above as I proposed.

Hah, well spotted.

OK, I'll change it to the old style. But, if this bitrots again...

rainers replied:
Thanks. No guarantees, though ;-)

rainers reviewed (commented):
LGTM but the mmixin
